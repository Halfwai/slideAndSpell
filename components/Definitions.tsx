import React, { useRef, useEffect, useState } from 'react';
import { View, Text, StyleSheet, Animated, Dimensions, DeviceEventEmitter, ScrollView } from 'react-native';
import Definition from '@/components/Definition';
import { COLOURS } from '@/constants/colours';

interface DefinitionsProps {
    slideIn: boolean,
    validWords: {
        word: string,
        definition: string
    }[]
}

export default function Definitions(props: DefinitionsProps) {
    const definitionsPositionY = useRef(new Animated.Value(Dimensions.get('window').height)).current;

    const wordList = useRef(props.validWords);

    useEffect(() => {
        wordList.current = props.validWords;
    }, [props.validWords]);

    useEffect(() => {
        if (props.slideIn) {
            Animated.timing(definitionsPositionY, {
                toValue: 0,
                duration: 500,
                useNativeDriver: true
            }).start();
        }
    }, [props.slideIn]);

    return (
        <Animated.View style={[styles.container, { transform: [{ translateX: 0 }, { translateY: definitionsPositionY }] }]}>
            <ScrollView>
                {wordList.current.map((word, index) => {
                    return (
                        <Definition key={index} word={word.word} definition={word.definition} />
                    )
                })}
            </ScrollView>
        </Animated.View>
    )
}

const styles = StyleSheet.create({
    container: {
        height: "80%",
        borderTopWidth: 1,
        borderBottomWidth: 1,
        borderColor: COLOURS.green,
        backgroundColor: COLOURS.green, 
        borderRadius: 10,
    }
});