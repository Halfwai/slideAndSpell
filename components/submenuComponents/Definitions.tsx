import React, { useRef, useEffect } from 'react';
import { StyleSheet, Animated, Dimensions, ScrollView } from 'react-native';
import Definition from '@/components/submenuComponents/Definition';
import { COLOURS } from '@/constants/colours';
import AuthButton from '@/components/buttons/AuthButton';

interface DefinitionsProps {
    slideIn: boolean,
    validWords: {
        word: string,
        definition: string
    }[],
    close?: Function
}

export default function Definitions(props: DefinitionsProps) {
    const definitionsPositionY = useRef(new Animated.Value(Dimensions.get('window').height)).current;
    const wordList = useRef(props.validWords);
    
    useEffect(() => {
        wordList.current = props.validWords;
    }, [props.validWords]);

    useEffect(() => {
        if (props.slideIn) {
            Animated.timing(definitionsPositionY, {
                toValue: 0,
                duration: 500,
                useNativeDriver: true
            }).start();
        }
    }, [props.slideIn]);

    function slideOut() {
        Animated.timing(definitionsPositionY, {
            toValue: Dimensions.get('window').height,
            duration: 500,
            useNativeDriver: true
        }).start(({ finished }) => {
            if (finished) props.close && props.close();
        });
    }

    return (
        <Animated.View style={[styles.container, { transform: [{ translateX: 0 }, { translateY: definitionsPositionY }] }]}>
            <ScrollView>
                {wordList.current.map((word, index) => {
                    return (
                        <Definition key={index} word={word.word} definition={word.definition} />
                    )
                })}
            </ScrollView>
            {props.close &&
                <AuthButton
                    text="Close"
                    onPress={() => slideOut()}
                    style={styles.button}
                />
            }
        </Animated.View>
    )
}

const styles = StyleSheet.create({
    container: {
        height: "100%",
        width: "100%",
        borderTopWidth: 1,
        borderBottomWidth: 1,
        borderColor: COLOURS.green,
        backgroundColor: COLOURS.green,
        borderRadius: 10,
        justifyContent: 'center',
        alignItems: 'center',
        position: 'absolute',
    },
    button: {
        backgroundColor: "white",
        marginVertical: 20,
    }
});