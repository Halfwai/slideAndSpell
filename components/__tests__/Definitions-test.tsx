import { render } from '@testing-library/react-native';

import Definitions from '@/components/Definitions';

it(`Definitions Component renders correctly`, () => {
    const { toJSON } = render(
        <Definitions 
            slideIn={true}
            validWords={[
                {
                    word: "word",
                    definition: "test definition"
                }
            ]}
        />
    );
    expect(toJSON()).toMatchSnapshot();
});