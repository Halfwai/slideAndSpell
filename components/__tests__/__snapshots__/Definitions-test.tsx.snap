// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Definitions Component renders correctly 1`] = `
<View
  collapsable={false}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#8BF0BC",
      "borderBottomWidth": 1,
      "borderColor": "#8BF0BC",
      "borderRadius": 10,
      "borderTopWidth": 1,
      "height": "100%",
      "justifyContent": "center",
      "position": "absolute",
      "transform": [
        {
          "translateX": 0,
        },
        {
          "translateY": 1334,
        },
      ],
      "width": "100%",
    }
  }
>
  <RCTScrollView>
    <View>
      <View
        style={
          {
            "backgroundColor": "white",
            "borderColor": "#8BF0BC",
            "borderRadius": 10,
            "borderWidth": 1,
            "margin": 10,
            "padding": 20,
            "width": 675,
          }
        }
      >
        <Text
          style={
            [
              {
                "fontFamily": "Quicksand",
              },
              undefined,
            ]
          }
        >
          <Text
            style={
              [
                {
                  "fontFamily": "Quicksand",
                },
                {
                  "fontSize": 18,
                  "fontStyle": "italic",
                  "textTransform": "capitalize",
                },
              ]
            }
          >
            word
            : 
          </Text>
          <Text
            style={
              [
                {
                  "fontFamily": "Quicksand",
                },
                {
                  "fontSize": 18,
                },
              ]
            }
          >
            test definition
          </Text>
        </Text>
        <Text
          href="/dictionary?word=word"
          onPress={[Function]}
          role="link"
          style={
            {
              "alignSelf": "flex-end",
              "borderRadius": 5,
              "borderWidth": 1,
              "marginTop": 10,
              "padding": 5,
            }
          }
        >
          <Text
            style={
              [
                {
                  "fontFamily": "Quicksand",
                },
                undefined,
              ]
            }
          >
            More info
          </Text>
        </Text>
      </View>
    </View>
  </RCTScrollView>
</View>
`;
